src/main/java/liquibase/ext/spanner/CloudSpanner.java

```diff
@@ -79,6 +83,41 @@ public String getDefaultDriver(String url) {
     }
     return null;
   }
+  
+  @Override
+  public void setConnection(final DatabaseConnection conn) {
+    DatabaseConnection connectionToUse = conn;
+    // If a user creates a JDBC connection manually and then passes this manually into Liquibase,
+    // then this method will be called by Liquibase. Normally, a connection will be opened by
+    // Liquibase based on the connection URL that is configured. In that case, the
+    // CloudSpanneConnection class will ensure that the correct user-agent string is set. That is
+    // not the case when a user creates the connection manually. This method therefore checks
+    // whether it is actually a Spanner JDBC connection, and if it is, replaces it with a new
+    // connection that uses the same connection URL + the user-agent string. The original connection
+    // is kept open in case the caller also uses the connection for other purposes, but will
+    // automatically be closed when the 'replacement' connection is closed. The latter should be
+    // safe, even if the caller uses the connection for other purposes, as even if the connection
+    // was not replaced it would have been closed by Liquibase at the same moment.
+    if (!(conn instanceof CloudSpannerConnection) && conn instanceof JdbcConnection
```

Comment:
> I assume that "manually creating the connection" refers to creating it through an API and using Liquibase that way -- as we do in the test harness against emulator/real Spanner.

Yes, correct. I've updated the comment to (hopefully) reflect that a little bit better.

> If someone were to use CloudSpannerJdbcConnection with shadow/shading I believe it could be renamed? So it may also not work. It wouldn't be the end of the world -- everything would still work correctly overall.

Yes, correct. This `if` check would in that case fail and the user agent string would not be set. It would however require that:
1. The application uses a shaded JAR with the JDBC driver.
2. AND that the application creates the connection programmatically instead of configuring it in the Liquibase configuration.